pkg/controller/service/service_controller.go

```diff
@@ -720,47 +617,104 @@ func (s *ServiceController) syncService(key string) error {
-	case err != nil:
-		glog.Infof("Unable to retrieve service %v from store: %v", key, err)
-	default:
-		cachedService = s.cache.getOrCreate(key)
-		err = s.processServiceUpdate(cachedService, service, key)
+	if errors.IsNotFound(err) {
+		glog.V(4).Infof("Service %q has been deleted", key)
+		return nil
 	}
 
-	return err
+	if err != nil {
+		glog.Errorf("Unable to retrieve service %v from store: %v", key, err)
+		return err
+	}
+
+	if isDeletionCandidate(service) {
+		err := s.processServiceDeletion(service, key)
+		if err != nil {
+			return err
+		}
+
+		service, err = s.removeFinalizer(service)
+		if err != nil {
+			return err
+		}
+
+		// The service and all external resources, if applicable, have been removed.
+		return nil
+	}
+
+	if needToAddFinalizer(service) {
+		service, err = s.addFinalizer(service)
+		if err != nil {
+			return err
+		}
+
+		// Continue processing the service now that it has been updated with the finalizer.
+	}
+
+	return s.processServiceUpdate(service, key)
 }
 
 // Returns an error if processing the service deletion failed, along with a time.Duration
 // indicating whether processing should be retried; zero means no-retry; otherwise
 // we should retry after that Duration.
-func (s *ServiceController) processServiceDeletion(key string) error {
-	cachedService, ok := s.cache.get(key)
-	if !ok {
-		glog.Errorf("service %s not in cache even though the watcher thought it was. Ignoring the deletion", key)
-		return nil
-	}
-	return s.processLoadBalancerDelete(cachedService, key)
+func (s *ServiceController) processServiceDeletion(service *v1.Service, key string) error {
+	return s.processLoadBalancerDelete(service, key)
 }
 
-func (s *ServiceController) processLoadBalancerDelete(cachedService *cachedService, key string) error {
-	service := cachedService.state
-	// delete load balancer info only if the service type is LoadBalancer
-	if !wantsLoadBalancer(service) {
-		return nil
-	}
+func (s *ServiceController) processLoadBalancerDelete(service *v1.Service, key string) error {
+	glog.Infof("Deleting existing load balancer for service %s that no longer needs a load balancer.", key)
 	s.eventRecorder.Event(service, v1.EventTypeNormal, "DeletingLoadBalancer", "Deleting load balancer")
 	err := s.balancer.EnsureLoadBalancerDeleted(context.TODO(), s.clusterName, service)
 	if err != nil {
 		s.eventRecorder.Eventf(service, v1.EventTypeWarning, "DeletingLoadBalancerFailed", "Error deleting load balancer (will retry): %v", err)
 		return err
 	}
+
 	s.eventRecorder.Event(service, v1.EventTypeNormal, "DeletedLoadBalancer", "Deleted load balancer")
-	s.cache.delete(key)
 
 	return nil
 }
+
+func (s *ServiceController) addFinalizer(svc *v1.Service) (*v1.Service, error) {
+	if hasFinalizer(svc) {
+		return svc, nil
+	}
+
+	updated := svc.DeepCopy()
+	updated.ObjectMeta.Finalizers = append(svc.ObjectMeta.Finalizers, serviceLoadBalancerFinalizer)
+
+	svc, err := serviceutil.PatchStatus(s.kubeClient.CoreV1(), svc, updated)
```

Comment:
Should this just be `Patch` and not `PatchStatus`?

Reply:
`PatchStatus` sounds better as the intention here is to only update status, right?

Reply:
Also adding a side note that was mentioned [before](https://github.com/kubernetes/kubernetes/pull/54569#pullrequestreview-97356951) ---  would be great to separate this patch change out so it won't be blocked by getting the finalizer KEP approved. Happy to help on sending a PR if you don't mind.